### PR TITLE
[Merged by Bors] - activation: publish atx based on local clock

### LINKED ISSUE #3774: ATX: remove restrictions on positioning ATX
## Description

currently positioning ATX must be from the previous epoch. this will prevent network from healing on its own when there are 1+ epochs without ATXs.

- positioning ATX can be a previous ATX of any smeshers from any previous epochs
- ATX need to declare explicitly its publish epoch (instead of inferring from its positioning ATX)

### LINKED ISSUE #4235: atxs: change expectations in the code that waits for the first atx
we have code that waits for first atx https://github.com/spacemeshos/go-spacemesh/blob/develop/activation/activation.go#L284-L306

the goal here is too select good enough atx from the recent epoch, so that new smesher is onboarded right for the next epoch. however we assume a bit small proving time, which will not allow us to select recent atx and will cause new smesher to wait one more additional epoch, before it will start getting rewards.

i think the good enough solution is to tie it to the cycle gap, lets say wait till 4/5 of cycle gap, and then select best available atx.

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/4235
closes: https://github.com/spacemeshos/go-spacemesh/issues/3774

in this change builder waits till the last point (poet round start - grace period) to get the most fresh and best positioning atx. it will try to submit challenge as long as there is some window before poet round starts, otherwise it will fail and will have to wait for another epoch. this code works the same way for initial atx and for long running smesher.

and, it changes how long we wait for poet to return proofs. grace period now used for two purposes:
- it gives poet time to prepare proof and serve Proof grpc request
- and it should be sufficient to prepare poet pow and submit grpc request with challenge

effective post time should fit into cycle gap - 2 * grace periods.

### CODE REVIEW

File: activation/activation_test.go [hunk 1]
```diff
@@ -173,11 +173,18 @@ func publishAtx(
...
 
 	publishEpoch := posEpoch + 1
-	tab.mclock.EXPECT().CurrentLayer().Return(*currLayer).AnyTimes()
+	// this approach makes no sense for mocking. it rewrites current layer for every
+	// subsequent request, and makes all future calls noop
+	if currentLayerExpect < 0 {
+		tab.mclock.EXPECT().CurrentLayer().Return(*currLayer).AnyTimes()
+	} else {
+		tab.mclock.EXPECT().CurrentLayer().Return(*currLayer).Times(currentLayerExpect)
+	}
```
Comment (fasmat) [inline activation/activation_test.go]: I feel like this could be called by the caller of `publishAtx`?

**Review by fasmat — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
@fasmat could you please review it again?

File: activation/activation.go [hunk 2]
```diff
@@ -345,22 +280,12 @@ func (b *Builder) loop(ctx context.Context) {
...
 		if err == nil {
-			b.log.WithContext(ctx).With().Info("waiting for atx to propagate before building the next challenge",
-				log.Duration("wait", b.poetRetryInterval),
-			)
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(b.poetRetryInterval):
-			}
 			continue
```
Comment (fasmat) [inline activation/activation.go:283]: I would not immediately retry again. We already had issues with nodes essentially DOS-ing on our poets and ending up in an endless loop without ever recovering.

File: activation/activation.go [hunk 3]
```diff
@@ -403,26 +328,53 @@ func (b *Builder) buildNIPostChallenge(ctx context.Context) (*types.NIPostChalle
...
 	}
+	current := b.currentEpoch()
+	prev, err := b.cdb.GetLastAtx(b.nodeID)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNotFound) {
+			return nil, err
+		}
+	} else if prev.PublishEpoch == current+1 {
+		current += 1
+	}
```
Comment (fasmat) [inline activation/activation.go:339]:
```suggestion
	if err != nil && !errors.Is(err, sql.ErrNotFound) {
		return nil, err
	}
	if prev.PublishEpoch == current+1 {
		current += 1
	}
```

Why is the check for `prev.PublishEpoch` necessary? Will the node eagerly create ATXs for more than one epoch in advance?

File: activation/activation.go [hunk 4]
```diff
@@ -403,26 +328,53 @@ func (b *Builder) buildNIPostChallenge(ctx context.Context) (*types.NIPostChalle
...
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(wait):
+		}
+	} else if until <= 0 {
+		metrics.PublishLateWindowLatency.Observe(-until.Seconds())
+		return nil, fmt.Errorf("%w: builder doesn't have time to submit in epoch %d. poet round already started %v ago",
+			ErrATXChallengeExpired, current, -until)
+	}
```
Comment (fasmat) [inline activation/activation.go]: I prefer checking for the error case first:

```suggestion
	until := time.Until(b.poetRoundStart(current))
	if until <= 0 {
		metrics.PublishLateWindowLatency.Observe(-until.Seconds())
		return nil, fmt.Errorf("%w: builder doesn't have time to submit in epoch %d. poet round already started %v ago",	ErrATXChallengeExpired, current, -until)
	}
	metrics.PublishOntimeWindowLatency.Observe(until.Seconds())
	if until > b.poetCfg.GracePeriod {
		wait := until - b.poetCfg.GracePeriod
		b.log.WithContext(ctx).With().Debug("waiting for fresh atxs",
			log.Duration("till poet round", until),
			log.Uint32("current epoch", current.Uint32()),
			log.Duration("wait", wait),
		)
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(wait):
		}
	}
```

File: activation/activation.go [hunk 5]
```diff
@@ -506,11 +458,13 @@ func (b *Builder) PublishActivationTx(ctx context.Context) error {
 
 	challenge, err := b.loadChallenge()
 	if err != nil {
-		logger.With().Info("challenge not loaded", log.Err(err))
-		logger.With().Info("building new atx challenge", log.Stringer("current_epoch", b.currentEpoch()))
+		logger.With().Info("building new atx challenge",
+			log.Stringer("current_epoch", b.currentEpoch()),
+			log.Err(err),
```
Comment (fasmat) [inline activation/activation.go:462]: does this error need to be logged? Won't it just be `sql.NotFound` most of the time?

File: activation/activation_test.go [hunk 6]
```diff
@@ -1113,3 +1078,51 @@ func TestBuilder_UpdatePoetsUnstable(t *testing.T) {
...
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			tab := newTestBuilder(t, WithPoetConfig(PoetConfig{
+				PhaseShift:  tc.shift,
+				GracePeriod: tc.grace,
+			}))
+			tab.mclock.EXPECT().CurrentLayer().Return(types.LayerID(0)).AnyTimes()
+			tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(func(lid types.LayerID) time.Time {
+				return genesis.Add(time.Duration(lid) * layerDuration)
+			}).AnyTimes()
```
Comment (fasmat) [inline activation/activation_test.go:1103]: NIT:

```suggestion
			tab.mclock.EXPECT().LayerToTime(gomock.Any()).Return(genesis.Add(time.Duration(lid) * layerDuration)).AnyTimes()
```

File: activation/activation.go (hunk 2 above)
Comment (dshulyak) [reply activation/activation.go:283]: this case is for err == nil. if we exited successfully - node submitted atx in the epoch and will wait for the next round to submit for the next one

File: activation/activation.go (hunk 3 above)
Comment (dshulyak) [reply activation/activation.go:339]: lets say node successfully submitted atx, and calls this function again (might be after restart). we check that it already submitted atx, and therefore needs to wait for +1 round

Comment (dshulyak) [reply activation/activation.go:339]: btw the suggestion doesn't work, as we can't call prev is error is not nill

File: activation/activation.go (hunk 5 above)
Comment (dshulyak) [reply activation/activation.go:462]: i wanted to keep it, just in case

File: activation/activation_test.go (hunk 6 above)
Comment (dshulyak) [reply activation/activation_test.go:1103]: this should return different results based on lid argument

**Comment by dshulyak** [comment]
bors try

File: activation/activation.go (hunk 3 above)
Comment (fasmat) [reply activation/activation.go:339]: Ah yes, you're right regarding the suggested code. 🙂 

But I'm still failing to understand the `prev.PublishEpoch == current+1` check?

If the prev. ATX is form the current epoch then all is fine and we create one for the next, but if prev. ATX is for the next epoch we create another one for the epoch after that?

**Comment by dshulyak** [comment]
bors try

File: activation/activation.go (hunk 3 above)
Comment (fasmat) [reply activation/activation.go:339]: 🤦 never mind we wait until `current`...

**Review by fasmat — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**